auth/src/main/java/com/firebase/ui/auth/ui/email/RegisterEmailActivity.java

```diff
@@ -135,11 +137,30 @@ public void onClick(View view) {
         });
     }
 
-    private void registerUser(String email, final String name, final String password) {
+    private void registerUser(String email, String name, String password) {
         final FirebaseAuth firebaseAuth = mActivityHelper.getFirebaseAuth();
         // create the user
-        firebaseAuth.createUserWithEmailAndPassword(email, password)
-                .addOnFailureListener(new TaskFailureLogger(TAG, "Error creating user"))
+        FirebaseUser user = firebaseAuth.getCurrentUser();
+        if (mActivityHelper.getFlowParams().shouldLinkUser && user != null) {
+            final String prevUid = user.getUid();
+            addListeners(user.linkWithCredential(EmailAuthProvider.getCredential(email, password))
+                                 .addOnSuccessListener(
+                                         new OnSuccessListener<AuthResult>() {
+                                             @Override
+                                             public void onSuccess(AuthResult result) {
+                                                 AuthUI.getInstance()
+                                                         .notifyOnMergeFailedListeners(prevUid);
+                                             }
+                                         }), name, password);
+        } else {
+            addListeners(firebaseAuth.createUserWithEmailAndPassword(email, password),
+                         name,
+                         password);
+        }
+    }
+
+    private void addListeners(Task<AuthResult> task, final String name, final String password) {
```

Comment:
IMO this is a confusing abstraction, why not just add the listeners directly as before?